source/h2_frames.c

```diff
@@ -332,714 +269,900 @@ void aws_h2_frame_encoder_clean_up(struct aws_h2_frame_encoder *encoder) {
-int aws_h2_frame_continuation_init(struct aws_h2_frame_continuation *frame, struct aws_allocator *allocator) {
-    (void)allocator;
+struct aws_h2_frame *aws_h2_frame_new_ping(
+    struct aws_allocator *allocator,
+    bool ack,
+    const uint8_t opaque_data[AWS_H2_PING_DATA_SIZE]) {
+
+    /* PING can be pre-encoded */
+    const uint8_t flags = ack ? AWS_H2_FRAME_F_ACK : 0;
+    const size_t payload_len = AWS_H2_PING_DATA_SIZE;
+    const uint32_t stream_id = 0;
+
+    struct aws_h2_frame_prebuilt *frame =
+        s_h2_frame_new_prebuilt(allocator, AWS_H2_FRAME_T_PING, stream_id, payload_len, flags);
+    if (!frame) {
+        return NULL;
+    }
 
-    AWS_ZERO_STRUCT(*frame);
-    frame->base.type = AWS_H2_FRAME_T_CONTINUATION;
+    /* Write the PING payload (RFC-7540 6.7):
+     * +---------------------------------------------------------------+
+     * |                                                               |
+     * |                      Opaque Data (64)                         |
+     * |                                                               |
+     * +---------------------------------------------------------------+
+     */
+    if (!aws_byte_buf_write(&frame->encoded_buf, opaque_data, AWS_H2_PING_DATA_SIZE)) {
+        aws_raise_error(AWS_ERROR_SHORT_BUFFER);
+        goto error;
+    }
 
-    return aws_h2_frame_header_block_init(&frame->header_block, allocator);
-}
-void aws_h2_frame_continuation_clean_up(struct aws_h2_frame_continuation *frame) {
-    AWS_PRECONDITION(frame);
+    return &frame->base;
 
-    aws_h2_frame_header_block_clean_up(&frame->header_block);
+error:
+    aws_h2_frame_destroy(&frame->base);
+    return NULL;
 }
 
-int aws_h2_frame_continuation_encode(
-    struct aws_h2_frame_continuation *frame,
-    struct aws_h2_frame_encoder *encoder,
-    struct aws_byte_buf *output) {
-    AWS_PRECONDITION(frame);
-    AWS_PRECONDITION(encoder);
-    AWS_PRECONDITION(output);
-
-    const size_t output_init_len = output->len;
-
-    /* Calculate length & flags */
-    size_t length = 0;
-    if (aws_h2_frame_header_block_get_encoded_length(&frame->header_block, encoder, &length)) {
-        goto compression_error;
+/***********************************************************************************************************************
+ * GOAWAY
+ **********************************************************************************************************************/
+static const size_t s_frame_goaway_length_min = 8;
+
+struct aws_h2_frame *aws_h2_frame_new_goaway(
+    struct aws_allocator *allocator,
+    uint32_t last_stream_id,
+    uint32_t error_code,
+    struct aws_byte_cursor debug_data) {
+
+    /* If debug_data is too long, don't sent it.
+     * It's more important that the GOAWAY frame gets sent. */
+    const size_t debug_data_max = s_prebuilt_payload_max() - s_frame_goaway_length_min;
+    if (debug_data.len > debug_data_max) {
```

Comment:
Would truncation be reasonable here? What would this payload look like?

Reply:
I think truncation is probably bad. Imagine owning the xml or json parser on the other side?

Reply:
Users are free to send whatever they want as the debug data.

Henso and I had a big chat about how to handle debug-data being too long.
We decided it shouldn't be a show-stopping error, it's more important that the GOAWAY frame gets sent.

We were less certain about truncating vs sending nothing when it's too long. Someone might never see their data get too long until the magic day that it does. So do we send partial data, or none at all? Partial XML or JSON would turn into a parse error, but a truncated log ... it might not be clear that it's been truncated and just be misleading. Options I can think of:
1) send no debug-data
1) truncate debug-data
1) truncate and write "TRUNCATED" as the last 9 bytes?

THoughts? We can always change this later. It's not exposed all the way out to a public API yet

Reply:
Jonathan chimed in below and voted again for no data

Reply:
I agree, thanks for clarifying.